integration/sawtooth_integration/tests/node_controller.py

```diff
@@ -0,0 +1,263 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# --------------------------------------------------------------------------
+
+import subprocess
+import shlex
+import time
+import logging
+
+from sawtooth_integration.tests.integration_tools import wait_for_rest_apis
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+# peering arrangements
+
+def peer_with_genesis_only(num):
+    if num > 0:
+        return '--peers {}'.format(
+            public_uri(0))
+    else:
+        return ''
+
+def peer_to_preceding_only(num):
+    if num > 0:
+        return '--peers {}'.format(
+            public_uri(num - 1))
+    else:
+        return ''
+
+def everyone_peers_with_everyone(num):
+    if num > 0:
+        peers = ','.join(public_uri(i) for i in range(num))
+        return '--peers {}'.format(peers)
+    else:
+        return ''
+
+
+# processor arrangements
+
+def intkey_config_registry(num):
+    # all nodes get the same processors
+    return 'tp_intkey_python', 'tp_config', 'tp_validator_registry'
+
+def intkey_xo_config_registry(num):
+    # all nodes get the same processors
+    return (
+        'tp_intkey_python',
+        'tp_xo_python',
+        'tp_config',
+        'tp_validator_registry'
+    )
+
+
+# node
+
+def start_node(num,
+               processor_func,
+               peering_func,
+               poet_kwargs):
+    rest_api = start_rest_api(num)
+    processors = start_processors(num, processor_func)
+    validator = start_validator(num, peering_func, poet_kwargs)
+
+    wait_for_rest_apis(['0.0.0.0:{}'.format(8080 + num)])
+
+    return [rest_api] + processors + [validator]
+
+def stop_node(process_list):
+    # This may seem a little dramatic, but seriously,
+    # validators don't go down without a fight.
+    for _ in range(2):
+        for process in process_list:
+            pid = process.pid
+            LOGGER.debug('Stopping process {}'.format(pid))
+            subprocess.run(shlex.split(
+                'kill -2 {}'.format(pid)))
```

Comment:
If you are using `Popen` to start these sub-processes, you can use `send_signal` to send a signal (`signal.SIGTERM`, etc.) to the sub-process.  And if you care, you can use `wait` with a reasonable timeout if you want to give the sub-process time to exit gracefully.